### PR TITLE
Add `map` and `when` extension methods over `StockResponse`

### PR DESCRIPTION
**Pull request by leynier**
## Description

The goal is to use the `StockResponse` in a more declarative way by using extension methods on `StockResponse`.

### Methods

- [x] `map` method
- [x] `maybeMap` method
- [x] `when` method
- [x] `maybeWhen` method

### Tests

- [x] `map` test
- [x] `maybeMap` test
- [x] `when` test
- [x] `maybeWhen` test

### Example

```dart
// Create a stream to listen tweet changes of the user `xmartlabs`
stock
    .stream('xmartlabs', refresh: true)
    .listen((StockResponse<List<Tweet>> stockResponse) {
  if (stockResponse is StockResponseLoading) {
    _displayLoadingIndicator();
  } else if (stockResponse is StockResponseData) {
    _displayTweetsInUI(stockResponse.requireData());
  } else {
    _displayErrorInUi((stockResponse as StockResponseError).error);
  }
});
```

With `when`:

```dart
// Create a stream to listen tweet changes of the user `xmartlabs`
stock
  .stream('xmartlabs', refresh: true)
  .listen((StockResponse<List<Tweet>> stockResponse) => stockResponse.when(
        onLoading: _displayLoadingIndicator,
        onData: _displayTweetsInUI,
        onError: (error, stackTrace) => _displayErrorInUi(error),
      ));
```

### CODE REVIEW

File: lib/src/stock_response_extensions.dart [hunk 1]
```diff
@@ -3,6 +3,67 @@ import 'package:stock/src/stock_response.dart';
...
+  }) =>
+      map(
+        onLoading: onLoading ?? (_) => orElse(),
+        onData: onData ?? (_) => orElse(),
+        onError: onError ?? (_) => orElse(),
+      );
+
+  /// Invoke [onData] if the response is successful, [onLoading] if the response
+  /// is loading, and [onError] if the response is an error.
+  E when<E>({
```
Comment (mirland) [inline lib/src/stock_response_extensions.dart:41]: What do you think of adding the origin? For example it could be useful to "stop the loading indicator"

**Review by mirland — COMMENTED** [review]
Thanks for the contribution! 
I think is easier using these extensions.
I left you just a small question

**Comment by mirland** [comment]
Hey @leynier, how are you doing? Could you check my comment? Ping me if you have some question/comment.

We’re planning to release the v1.0 and it would be great to include your feature!

File: lib/src/stock_response_extensions.dart (hunk 1 above)
Comment (leynier) [reply lib/src/stock_response_extensions.dart:41]: Yeah, that would be useful.

**Comment by leynier** [comment]
Hi @mirland, sorry for the late response, I had a blackout for several days due to hurricane Ian.

I could have the tests and the origin parameter addition ready by early next week.

Best regards.

**Comment by mirland** [comment]
Cool, that great! Again thanks for your contribution!

**Comment by leynier** [comment]
@mirland ready for review ✅

**Review by f7deleon — APPROVED** [review]
Thanks for contributing, looks good to me!

**Review by mirland — APPROVED** [review]
THanks!

**Pull request merged**